### DISCUSSION TITLE
Options for remote URL images in Markdown coming from headless CMS via JavaScript Data Files

### DISCUSSION THREAD

**Discussion by antgel** [solution-loisir/markdown-it-eleventy-img#59] [Q&A]
Following on from the [comment that I piggy-backed on Issue #2](https://github.com/solution-loisir/markdown-it-eleventy-img/issues/2#issuecomment-1700275131):

Firstly @solution-loisir I'd like to thank you for being responsive and for the thorough answer.

I have some other user stories to work on first, but I will have a go at importing the remote sources at some point. Build time is not much of a concern as the sites are _relatively_ small and not updated too often.

About the idea of the Nunjucks shortcode, let me get into some more detail.

In the current workflow, the users can upload an image to Strapi's Media Library (backed by S3 / CloudFront) in the Strapi admin panel. Then they write their blog post in a [Strapi Rich Text](https://docs.strapi.io/user-docs/content-manager/writing-content) field, click on the image icon in the editor, then select the image, and the editor inserts the relevant Markdown.

So they don't really write Markdown as such, and nor do I think that would be suitable. They are intelligent, but they are "normal" people.

Then, Strapi sends the entire blog post content in a single Markdown field `content` which is currently displayed in my Nunjucks template with something like:
`<div>{{ post.attributes.content | markdown | safe }}</div>`

For reference, `markdown` is defined as:
```javascript
  const md = new markdownIt();
  md.use(markdownItAttrs);
  eleventyConfig.addFilter("markdown", (content) => {
    return md.render(content);
  });
```

If my Markdown contains something like:
```markdown
![Cat playing](https://assets.prod.web.example.com/cat_playing.jpg){.img-fluid}
```
this is transformed into the correct HTML, including the `img-fluid` class that Bootstrap needs.

However, I tried adding some Nunjucks code into the same `content` field with `templateEngineOverride: njk,md` in my 11ty front matter, and it didn't like it in the slightest. For example, I added:
```nunjucks
{% image "https://assets.prod.web.example.com/cat_playing.jpg", "Cat playing", "img-fluid", "(max-width: 500px) 100vw, (max-width: 750px) 100vw, (max-width: 1000px) 100vw" %}
```

Where my `image` shortcode is defined as:
```javascript
  eleventyConfig.addShortcode("image", async function(src, alt, klass, sizes, loading) {
    let metadata = await Image(src, {
      widths: [500, 750, 1000, "auto"],  // aligned with Strapi size optimization
      formats: ["avif", "webp", "jpeg"],
      outputDir: "dist/img"
    });

    let imageAttributes = {
      class: klass,
      alt,
      sizes,
      loading: loading || "eager",
      decoding: "async",
    };
  });
```

However, this doesn't output the HTML I'd expect, in fact it's a complete mess:
```html
<p %="" image="" https:assets.prod.web.example.comcat_playing.jpg,="" catplaying,="" img-fluid,="" (max-width:500px)100vw,(max-width:750px)100vw,(max-width:1000px)100vw=""></p>
```

It doesn't make any difference if I do this with or without `templateEngineOverride: njk,md`. I suspect if this is meant to work, it may be to do with my `markdown` filter.

Even if I take Strapi out of the equation, it doesn't work. e.g. If for test purposes I edit my Nunjucks template to be something like:
```nunjucks
{# This isn't coming from the markdown field in Strapi any more! #}
{% image "https://assets.prod.web.example.com/cat_playing.jpg", "Cat playing", "img-fluid", "(max-width: 500px) 100vw, (max-width: 750px) 100vw, (max-width: 1000px) 100vw" %}

<div>{{ post.attributes.content | markdown | safe }}</div>
```
my `image` shortcode returns `undefined` instead of HTML.

I'm not sure how to debug this but I think I have a number of problems. Hope you can understand what I wrote and look forward to the discussion. :)

**Comment by solution-loisir** [comment]
Hmm, there's a lot going on here and even though it doesn't seem to be directly related to the markdown-it-eleventy-img plugin, I'll try my best to give you my opinion if it can help.

I noticed that your shortcode definition is not returning any HTML (or anything at all for that matter). I guess that's why you're getting `undefined`. Try the following to see if it works:
```js
eleventyConfig.addAsyncShortcode("image", async function(src, alt, klass, sizes, loading) {
    let metadata = await Image(src, {
      widths: [500, 750, 1000, "auto"],  // aligned with Strapi size optimization
      formats: ["avif", "webp", "jpeg"],
      outputDir: "dist/img"
    });

    let imageAttributes = {
      class: klass,
      alt,
      sizes,
      loading: loading || "eager",
      decoding: "async",
    };

    return Image.generateHTML(metadata, imageAttributes, {
      whitespaceMode: "inline"
    });
});
```
The `generateHTML` method is already packaged with the [eleventy-img] plugin. The `whitespaceMode: "inline"` property prevents the injection of a bunch of `<p>` tags after every line breaks by the markdown parser. It basically produces the result as a single line. 

Hopefully that helps. :grin:

**Comment by solution-loisir** [comment]
Also, I think that inserting the shortcode at the Strapi level is likely to not work since Strapi is not aware of such concept. Your best bet ~would~ could be to swap the image token for a shortcode in eleventy JavaScript data file.

**Comment by solution-loisir** [comment]
Feel free to reopen if it's still relevant. Thanks!
